### PR TITLE
Hashjoin(InnerJoin) performance opt by reducing virtual function calls

### PR DESCRIPTION
**Pull request by awakeljw**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Inner Join performance Improvement by reducing virtual function calls.
1. merge blocks to one in addJoinedBlock if memory is enough.
2. use insertIndicesFrom to insert a batch of rows instead of inserting one by one.

test case:
```
create table t1(a int, b int)engine=MergeTree order by a;
create table t2(a int, b int)engine=MergeTree order by a;
create table t3(a int, b int)engine=MergeTree order by a;
insert into t1 select number, number+1 from numbers(100000000);
insert into t2 select number, number+1 from numbers(20000000);
insert into t3 select number, number+1 from numbers(10000000);
 :) select count(*) from t1 inner join t2 on t1.a=t2.a inner join t3 on t2.b=t3.b;

SELECT count(*)
FROM t1
INNER JOIN t2 ON t1.a = t2.a
INNER JOIN t3 ON t2.b = t3.b

Query id: a3048709-0c92-4b05-b2b7-a75548d23202

┌──count()─┐
│ 10000000 │
└──────────┘

1 rows in set. Elapsed: 9.429 sec. Processed 130.00 million rows, 600.00 MB (13.79 million rows/s., 63.63 MB/s.)

:) set max_merged_block_size=102400000000;
:) select count(*) from t1 inner join t2 on t1.a=t2.a inner join t3 on t2.b=t3.b;

SELECT count(*)
FROM t1
INNER JOIN t2 ON t1.a = t2.a
INNER JOIN t3 ON t2.b = t3.b

Query id: ec116ed3-76cb-4e2d-bfcb-aea58861fafa

┌──count()─┐
│ 10000000 │
└──────────┘

1 rows in set. Elapsed: 5.928 sec. Processed 130.00 million rows, 600.00 MB (21.93 million rows/s., 101.21 MB/s.)

```
SSB test SF=100
<!DOCTYPE html>

SSB_Join | After opt(ms) | Before opt(ms)
-- | -- | --
Q1.1 | 232 | 230
Q1.2 | 198 | 197
Q1.3 | 205 | 199
Q2.1 | 1358 | 4073
Q2.2 | 1046 | 1162
Q2.3 | 996 | 1013
Q3.1 | 2022 | 17285
Q3.2 | 1239 | 4013
Q3.3 | 988 | 1140
Q3.4 | 385 | 396
Q4.1 | 2659 | 19693
Q4.2 | 1268 | 4254
Q4.3 | 1063 | 4210

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -346,6 +346,7 @@ class IColumn;
     \
     M(UInt64, max_rows_in_join, 0, "Maximum size of the hash table for JOIN (in number of rows).", 0) \
     M(UInt64, max_bytes_in_join, 0, "Maximum size of the hash table for JOIN (in number of bytes in memory).", 0) \
+    M(UInt64, max_merged_block_size, 0, "Maximum size of the merged block for JOIN (in number of bytes in memory).", 0) \
```
Comment (vdimir) [inline src/Core/Settings.h:349]: Isn't it the same as `max_joined_block_size_rows` (but it bytes not rows)?

**Comment by vdimir** [comment]
> Just curious, does this pr conflict with that one?: https://github.com/ClickHouse/ClickHouse/pull/36415

That PR implemented on top of `HashJoin` and don't change any internals, so it is independent.

File: src/Core/Settings.h (hunk 1 above)
Comment (awakeljw) [reply src/Core/Settings.h:349]: I think we can limit the bytes to avoiding memory overflow， but rows can not do that.

**Comment by awakeljw** [comment]
> > Just curious, does this pr conflict with that one?: #36415
> 
> That PR implemented on top of `HashJoin` and don't change any internals, so it is independent.
It should have some conflict in two PRs if two parameter validate at the same time. I will fix it.

**Comment by awakeljw** [comment]
@Mergifyio update

**Comment by kitaisreal** [comment]
@awakeljw please take a look at performance regression https://s3.amazonaws.com/clickhouse-test-reports/36782/2548812a4e390bbc12c33dc0a4b244e610e9923c/performance_comparison__actions__[4/4]/report.html.

**Comment by awakeljw** [comment]
> @awakeljw please take a look at performance regression [https://s3.amazonaws.com/clickhouse-test-reports/36782/2548812a4e390bbc12c33dc0a4b244e610e9923c/performance_comparison__actions__[4/4]/report.html](https://s3.amazonaws.com/clickhouse-test-reports/36782/2548812a4e390bbc12c33dc0a4b244e610e9923c/performance_comparison__actions__%5B4/4%5D/report.html).

I checked framegraph and found a lot memcpy cost. Because I used more std::pair first and reconstruct to prepared_columns(std::vector<const IColumn *>) and row_nums(std::vector<size_t>). So I change to stuct MatchInfo, I hope it helps.

**Comment by awakeljw** [comment]
@kitaisreal I have submitted this pr for a while, can you help me improve my code and make it qualified.

**Pull request closed without merging**